File: leetcode/0051-n-queens/solution.cpp
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        this->n = n;
        colPlacement.assign(n, -1);
        backtrack(0);
        return result;
    }

private:
    int n;
    vector<int> colPlacement;
    unordered_set<int> usedCols, usedDiag1, usedDiag2;
    vector<vector<string>> result;

    void backtrack(int row) {
        if (row == n) {
            result.push_back(buildBoard());
            return;
        }

        for (int col = 0; col < n; col++) {
            if (usedCols.count(col) || usedDiag1.count(row - col) || usedDiag2.count(row + col)) {
                continue;
            }

            colPlacement[row] = col;
            usedCols.insert(col);
            usedDiag1.insert(row - col);
            usedDiag2.insert(row + col);

            backtrack(row + 1);

            usedCols.erase(col);
            usedDiag1.erase(row - col);
            usedDiag2.erase(row + col);
        }
    }

    vector<string> buildBoard() {
        vector<string> board;
        for (int r = 0; r < n; r++) {
            string line(n, '.');
            line[colPlacement[r]] = 'Q';
            board.push_back(line);
        }
        return board;
    }
};

File: leetcode/0051-n-queens/Solution.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    private int n;
    private int[] colPlacement;
    private Set<Integer> usedCols = new HashSet<>();
    private Set<Integer> usedDiag1 = new HashSet<>();
    private Set<Integer> usedDiag2 = new HashSet<>();
    private List<List<String>> result = new ArrayList<>();

    public List<List<String>> solveNQueens(int n) {
        this.n = n;
        this.colPlacement = new int[n];
        backtrack(0);
        return result;
    }

    private void backtrack(int row) {
        if (row == n) {
            result.add(buildBoard());
            return;
        }

        for (int col = 0; col < n; col++) {
            if (usedCols.contains(col) || usedDiag1.contains(row - col) || usedDiag2.contains(row + col)) {
                continue;
            }

            colPlacement[row] = col;
            usedCols.add(col);
            usedDiag1.add(row - col);
            usedDiag2.add(row + col);

            backtrack(row + 1);

            usedCols.remove(col);
            usedDiag1.remove(row - col);
            usedDiag2.remove(row + col);
        }
    }

    private List<String> buildBoard() {
        List<String> board = new ArrayList<>();
        for (int r = 0; r < n; r++) {
            char[] line = new char[n];
            java.util.Arrays.fill(line, '.');
            line[colPlacement[r]] = 'Q';
            board.add(new String(line));
        }
        return board;
    }
}
